Introduce abicoder pragma. (#10096)
ethereum/solidity: docs/layout-of-source-files.rst

  required by the pragma. If it does not match, the compiler issues
  an error.

ABI Coder Pragma
----------------

> Using ``pragma abicoder v1`` or ``pragma abicoder v2`` you can

Reviewer: ```suggestion
By using ``pragma abicoder v1`` or ``pragma abicoder v2`` you can
```
? the current version sounds a bit weird to me

select between the two implementations of the ABI encoder and decoder.

> The new ABI encoder (v2) is able to encode and decode arbitrarily nested

Reviewer: ```suggestion
The new ABI coder (v2) is able to encode and decode arbitrarily nested
```

arrays and structs. It might produce less optimal code and has not
received as much testing as the old encoder, but is considered
non-experimental as of Solidity 0.6.0. You still have to explicitly
activate it using ``pragma abicoder v2;``. Since it will be
activated by default starting from Solidity 0.8.0, there is the option to select
the old coder using ``pragma abicoder v1;``.

The set of types supported by the new encoder is a strict superset of
the ones supported by the old one. Contracts that use it can interact with ones
that do not without limitations. The reverse is possible only as long as the
non-``abicoder v2`` contract does not try to make calls that would require
decoding types only supported by the new encoder. The compiler can detect this
and will issue an error. Simply enabling ``abicoder v2`` for your contract is
enough to make the error go away.

.. note::
  This pragma applies to all the code defined in the file where it is activated,
  regardless of where that code ends up eventually. This means that a contract

>   whose source file is selected to compile with abi coder v1

Reviewer 2: ```suggestion
  whose source file is selected to compile with ABI coder v1
```

  can still contain code that uses the new encoder
  by inheriting it from another contract. This is allowed if the new types are only
  used internally and not in external function signatures.

.. note::

>   Up to Solidity 0.7.4, it was possible to select the ABI coder V2

Reviewer 2: ```suggestion
  Up to Solidity 0.7.4, it was possible to select the ABI coder v2
```

Should decide on a consistent naming in this file really :)

>   using ``pragma experimental ABIEncoderV2``, but it was not possible

Reviewer 2: ```suggestion
  by using ``pragma experimental ABIEncoderV2``, but it was not possible
```

  to explicitly select coder v1 because it was the default.

.. index:: ! pragma, experimental

.. _experimental_pragma:

Experimental Pragma
-------------------

The second pragma is the experimental pragma. It can be used to enable
features of the compiler or language that are not yet enabled by default.
The following experimental pragmas are currently supported:

ABIEncoderV2
~~~~~~~~~~~~

> Because the ABI coder version 2 is not considered experimental anymore,

Reviewer 2: ```suggestion
Because the ABI coder v2 is not considered experimental anymore,
```
